File: leetcode/100-999/160. Intersection of Two Linked Lists/solution_1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution 
{
    public:
        ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) 
        {
            auto A = headA;
            auto B = headB;

            while (A && B)
            {
                if (A == B)
                    return A;
                else
                {
                    if (A->next == nullptr)
                    {
                        A = headA;
                        B = B->next;
                    }
                    else
                        A = A->next;
                }
            }    

            return nullptr;
        }
};

File: leetcode/100-999/160. Intersection of Two Linked Lists/solution_1.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */

public class Solution 
{
    public ListNode GetIntersectionNode(ListNode headA, ListNode headB) 
    {
        var A = headA;
        var B = headB;

        while (A != null && B != null)
        {
            if (A == B)
                return A;
            else
            {
                if (A.next == null)
                {
                    A = headA;
                    B = B.next;
                }
                else
                    A = A.next;
            }
        }

        return null;
    }
}